mm2src/coins/eth/web3_transport/http_transport.rs

```diff
@@ -194,129 +124,110 @@ async fn send_request(
+            request,
+            Web3RpcError::Transport(format!(
                 "Server: '{}', response !200: {}, {}",
                 node.uri,
                 status,
                 binprint(&body, b'.')
-            )));
-            continue;
-        }
-
-        let res = match single_response(body, &node.uri.to_string()) {
-            Ok(r) => r,
-            Err(err) => {
-                errors.push(Web3RpcError::InvalidResponse(format!(
-                    "Server: '{}', error: {}",
-                    node.uri, err
-                )));
-                continue;
-            },
-        };
-
-        client_impl.nodes.rotate_left(i);
-
-        return Ok(res);
+            )),
+        ));
     }
 
-    Err(request_failed_error(&request, &errors))
+    let res = match de_rpc_response(body, &node.uri.to_string()) {
+        Ok(r) => r,
+        Err(err) => {
+            return Err(request_failed_error(
+                request,
+                Web3RpcError::InvalidResponse(format!("Server: '{}', error: {}", node.uri, err)),
+            ));
+        },
+    };
+
+    Ok(res)
 }
 
 #[cfg(target_arch = "wasm32")]
 async fn send_request(
     request: Call,
-    client: Arc<HttpTransportRpcClient>,
+    node: HttpTransportNode,
     event_handlers: Vec<RpcTransportEventHandlerShared>,
     gui_auth_validation_generator: Option<GuiAuthValidationGenerator>,
 ) -> Result<Json, Error> {
-    let serialized_request = to_string(&request);
-
-    let mut errors = Vec::new();
-    let mut client_impl = client.0.lock().await;
-
-    for (i, node) in client_impl.nodes.clone().iter().enumerate() {
-        let serialized_request =
-            match handle_gui_auth_payload_if_activated(&gui_auth_validation_generator, node, &request) {
-                Ok(Some(r)) => r,
-                Ok(None) => serialized_request.clone(),
-                Err(e) => {
-                    errors.push(e);
-                    continue;
-                },
-            };
-
-        match send_request_once(serialized_request, &node.uri, &event_handlers).await {
-            Ok(response_json) => {
-                client_impl.nodes.rotate_left(i);
-                return Ok(response_json);
+    let mut serialized_request = to_string(&request);
+
+    if node.gui_auth {
+        match handle_gui_auth_payload(&gui_auth_validation_generator, &request) {
+            Ok(r) => serialized_request = r,
+            Err(e) => {
+                return Err(request_failed_error(
+                    request,
+                    Web3RpcError::Transport(format!("Server: '{}', error: {}", node.uri, e)),
+                ));
             },
-            Err(Error::Transport(e)) => {
-                errors.push(Web3RpcError::Transport(format!("Server: '{}', error: {}", node.uri, e)))
-            },
-            Err(e) => errors.push(Web3RpcError::InvalidResponse(format!(
-                "Server: '{}', error: {}",
-                node.uri, e
-            ))),
```

Comment:
There was some advantages to the old implementation. We used to try the next node if there was a response error like the one below as it was fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1998
```
25 23:41:28, coins::eth:4297] ERROR Error RPC error: Error { code: ServerError(-32001), message: "Exceeded the quota usage", data: None } on eth_gasPrice request
```
We also use `web3_clientversion` for each request to get a live client which is an additional overhead and cost on API Credits.

Reply:
I agree, will refactor this.

Reply:
Done

Reply:
Wouldn't this https://github.com/KomodoPlatform/komodo-defi-framework/blob/aab065eb3ac91fd49d0b5b79813a561e8e6feeec/mm2src/coins/eth.rs#L2545-L2549 just return the same client that we got the first time over and over, I guess you did it right before this simplification https://github.com/KomodoPlatform/komodo-defi-framework/pull/2058/commits/751ba29c380668bdb37f1b2b9194184ff1af55ab maybe you wanted to use `set_last_request_failed` inside `request_failed_error` but forgot.

Reply:
Fixed

Reply:
Last attempt should cover all the cases.